Add option enabled for each camera in config

banthungprong opened the pull request:
Based on requirement and discussion in #4109:
- for each camera an option "enabled: True/False" can be used
- if option isn't used in config file the camera is enable as default (so no changes in the old config are necessary

Corrected 2 typos in config.py (propagate, generate)

Adding the if block resulted in 91 deletions and 96 additions after changing the indentation. The only real changes were:
- in config.py in class CameraConfig
`enabled: Optional[bool] = Field(default=True, title="Enable camera.")`

- in config.py in class FrigateConfig in runtime_config:
replace 
`for name, camera in config.cameras.items():`
with 
`for name, camera in config.cameras.copy().items():`
... generate camera_config ...
`if camera_config.enabled:`
... configurations ...
`else: config.cameras.pop(name)`
befor line
`return config`

banthungprong committed:
add option enabled for each camera in config

NickM-27 commented on frigate/config.py:
```diff
@@ -816,109 +817,113 @@ def runtime_config(self) -> FrigateConfig:
         )
 
-        for name, camera in config.cameras.items():
+        for name, camera in config.cameras.copy().items():
```
Why was copy added?

NickM-27 replied:
Nevermind I get it. I think this would be easier to just add a filter instead of adding if / else.

banthungprong replied:
If you use it without, you'll get an error when modifying a lilst while iterating it. With the "copy trick" this can be avoided.

NickM-27 replied:
See my second comment. Another option besides using `filter` to avoid the large if / else could be doing 

```
if not camera.enabled:
  config.cameras.pop(camera)
  continue
```

banthungprong replied:
When I switch to this option I get an error
"TypeError: unhashable type: CameraConfig"
Just trying to understand why...

NickM-27 replied:
Which option (filter or the code snippet I put)? 

To be clear you'll still need the copy() with the code snippet I put above, it'll help clean up the large if / else that we'd prefer to avoid

NickM-27 commented on frigate/config.py:
```diff
@@ -530,6 +530,7 @@ class CameraUiConfig(FrigateBaseModel):
 
 class CameraConfig(FrigateBaseModel):
     name: Optional[str] = Field(title="Camera name.", regex="^[a-zA-Z0-9_-]+$")
+    enabled: Optional[bool] = Field(default=True, title="Enable camera.")
```
This isn't optional, it always needs a value of true or false

banthungprong replied:
I tested the different options in the config (with two cameras A and B):
- no entry "enabled: ..." --> both cameras are shown and no error is thrown (I assumed this is a result of the optional Bool setting)
- entry "enabled: False" for Camera A or B --> Camera A or B is disabled
- entry "enabled: True" for Camera A or B --> Camera A or B is enabled.

I do not really understand what might be wrong here.

NickM-27 replied:
You are using the `Optional` which means the value is nullable, but it shouldn't be. This will always have a default value of True or a user set value.

NickM-27 commented:
Also need to update configuration docs to reflect this new option

banthungprong commented:
Docs Update will follow if the reviews here are ok. I don't want to update the doc before the solution seems to be ok.
I can add the documenation update as a commit to this branch later or not?
If it is better to change something in code AND in doc at the same time I will do from now on.

NickM-27 commented:
Yes, docs and code updates should always be part of the same PR

NickM-27 commented on frigate/config.py:
```diff
@@ -530,6 +530,7 @@ class CameraUiConfig(FrigateBaseModel):
 
 class CameraConfig(FrigateBaseModel):
     name: Optional[str] = Field(title="Camera name.", regex="^[a-zA-Z0-9_-]+$")
+    enabled: Optional[bool] = Field(default=True, title="Enable camera.")
```
```suggestion
    enabled: bool = Field(default=True, title="Enable camera.")
```

You are using the Optional qualifier which means the field is nullable, but it shouldn't be. This will always have a default value of True or a user set value.

banthungprong replied:
Ok. Simply misunderstood the Optional.

banthungprong committed:
Simplified If-block and removed wrong Optional

banthungprong committed:
Update Docs enabling/disabling camera in config

banthungprong commented:
Code changes as requested and updated documentation for the new option

NickM-27 commented on docs/docs/configuration/index.md:
```diff
@@ -380,6 +380,8 @@ timestamp_style:
 cameras:
   # Required: name of the camera
   back:
+    # Optional: Enable/Disable the camera (default: camera is enabled automatically)
```
```suggestion
    # Optional: Enable/Disable the camera (default: shown below)
```

should fit the format of other options

NickM-27 reviewed (changes requested):
Looks good, just one small thing

banthungprong committed:
correct format for option

blakeblackshear commented on frigate/config.py:
```diff
@@ -816,12 +817,16 @@ def runtime_config(self) -> FrigateConfig:
             )
 
+            if not camera_config.enabled:
+                config.cameras.pop(name)
```
I haven't tested this, but won't this essentially just prevent the camera's config from being loaded at all? I thought the goal was to make it possible to have retain settings for a disabled camera that are different than the default values.

NickM-27 replied:
Yeah you're right, I wasn't sure if that was supposed to be in this initial pass. 

In that case we'd want to leave the config itself alone and not run the ffmpeg commands if the camera is disabled.

banthungprong replied:
I think we talk about 3 different requirements:
1. static: Option in config, only used one time when starting the container (the original requirement for this PR). Camera settings aren't loaded in config.py.
2. dynamic: Camera can be disabled or enabled while running frigate, so the config is changed in the UI. State of the camera will be saved "somewhere else" but not in the config. Settings are always available.
3. "A bit of both": Option in config, when starting frigate the settings are loaded but the views (recording/events/live/rtmp etc) do not show a disabled camera. But additionaly features in the UI to enable/disable every camera. Setting are always available.

Is it enough to "not run the ffmpeg commands" or is more logic necessary in the different views when a camera isn't shown (example: Events View, dropdown Filter)? I will start digging and trying here today. 

Close this pull request and make a new one based on requirement 2 or 3?
Or update this one and switch to the dynamic solution?

If starting to change parameters dynamically (enabled/disabled is only a first possibility) shouldn't there be a view for configurations showing each camera with a list of parameters that might be changed in the UI (resolution, framerate, etc.)?

NickM-27 replied:
The dynamic solution should come later, that's not what's being suggested. 

This solution should still be static. The thing that's being discussed is we don't want the cameras config completely removed as we still want the retention settings to be applied. If a camera is disabled it just shouldn't be running, it should still show in the dropdowns and other places because you may still want to view those events

banthungprong replied:
Ok. But this is a little bit difficult to understand:
If I disable a camera in the top of the camera config I wouldn't expect parts of the camera still "working" in the UI.

Does "not running" mean the following? 
- no motion, detect, recording, snapshot
- no live view and no debug view
- everything else would still be visible and usable: viewing events, snapshots and records.

I'll admit I have no idea how to solve this right now.

NickM-27 replied:
> Does "not running" mean the following?
> 
> * no motion, detect, recording, snapshot
> * no live view and no debug view
> * everything else would still be visible and usable: viewing events, snapshots and records.

Yes preciesely, disabled will mean that the ffmpeg / video processes won't run but the camera will otherwise still be in the UI. It should just consist of not starting a process around https://github.com/blakeblackshear/frigate/blob/dev/frigate/app.py#L250 and probably some other places as well

banthungprong replied:
Thanks! I started looking already in app.py. And "some other places". I will try it and I hope it won't get too annoying if my commits aren't perfect...

banthungprong replied:
Not starting the process when the camera is disabled is working already.
Birdseye, Events, Debug all ok.
In Cameras: I hate the green image when no camera snapshot/stream is available...

NickM-27 replied:
> In Cameras: I hate the green image when no camera snapshot/stream is available...

That is on the list to improve in the 0.12.0 release, so I'd say leave it with the green screens for now as that is a separate PR which will address the green screen. Maybe having one message for cameras with errors and a different message for disabled cameras.

banthungprong replied:
I do not start camera processors and capture processes when a camera is disabled. Working. Will commit in the morning. Too tired now.

Naming looks strange:
start_camera_proces**sors** vs start_capture_proces**ses**
whereas inside camera_process and capture_process looks ok.

banthungprong replied:
> That is on the list to improve in the 0.12.0 release, so I'd say leave it with the green screens for now as that is a separate PR which will address the green screen. Maybe having one message for cameras with errors and a different message for disabled cameras.

I couldn't resist:
![Screenshot from 2022-10-27 10-39-01](https://user-images.githubusercontent.com/59219161/198235486-6fe839bf-f9c3-4331-81a0-79f00b45f57e.png)

banthungprong committed:
Disabling Camera for processes, no config changes

banthungprong committed:
Describe effects of disabled cam in documentation

NickM-27 commented on frigate/app.py:
```diff
@@ -250,36 +250,42 @@ def start_video_output_processor(self) -> None:
     def start_camera_processors(self) -> None:
         model_shape = (self.config.model.height, self.config.model.width)
         for name, config in self.config.cameras.items():
```
Similar to previously, instead of the if / else we can just add

```
if not enabled:
  Log
  continue

```

banthungprong replied:
Will change. I simply do not like the historic "basic" constructs, with goto, break, continue. Old discussions about readability, workflow etc.

NickM-27 commented on docs/docs/configuration/index.md:
```diff
@@ -380,6 +380,10 @@ timestamp_style:
   # Required: name of the camera
   back:
+    # Optional: Enable/Disable the camera (default: shown below).
+    #           If disabled: config is used but no live stream and no capture etc.
```
Indentation is incorrect for these (# is correct but doesn't need space after)

NickM-27 commented on docs/docs/guides/getting_started.md:
```diff
@@ -62,6 +62,7 @@ detectors:
 
 cameras:
   camera_1: # <------ Name the camera
+    enabled: True # <----- Enable/disable camera
```
I'm not sure if it makes sense to put it here, this section is the minimal config and this isn't necessary

NickM-27 commented on frigate/app.py:
```diff
@@ -250,36 +250,42 @@ def start_video_output_processor(self) -> None:
+                logger.info(f"Camera processor not started for disabled camera {name}")
 
     def start_camera_capture_processes(self) -> None:
         for name, config in self.config.cameras.items():
```
Same here as above

NickM-27 commented on frigate/config.py:
```diff
@@ -816,7 +817,7 @@ def runtime_config(self) -> FrigateConfig:
         )
 
-        for name, camera in config.cameras.items():
+        for name, camera in config.cameras.copy().items():
```
Don't need the copy here anymore

NickM-27 reviewed (changes requested):
Just some small stuff

banthungprong committed:
change if-logic, obsolete copy, info disabled cam

banthungprong commented:
Commit comment length was limited (I just changed that in VS Code). So here a summary of the changes:
- changed the if-logic as reqested (2x) in app.py
- removed obsolete copy() in config.py
- removed enabled option in minimal camera configuration in getting_started.md
- added info in Cameras View when a camera is disabled and avoid green image in CameraImage.jsx

NickM-27 commented on docs/docs/configuration/index.md:
```diff
@@ -380,6 +380,10 @@ timestamp_style:
 cameras:
   # Required: name of the camera
   back:
+    # Optional: Enable/Disable the camera (default: shown below).
+    #           If disabled: config is used but no live stream and no capture etc.
+    #           Events/Recordings are still viewable.
```
```suggestion
    # Optional: Enable/Disable the camera (default: shown below).
    # If disabled: config is used but no live stream and no capture etc.
    # Events/Recordings are still viewable.
```

this is what I mean, the start of the comment is indented too much

NickM-27 replied:
this still needs to be addressed

banthungprong replied:
a few lines above after thickness: 2 the optional block (and more of these above) is formatted in the same way. But I changed it anyway.

NickM-27 commented on web/src/components/CameraImage.jsx:
```diff
@@ -45,12 +46,18 @@ export default function CameraImage({ camera, onload, searchParams = '', stretch
+      {
+        (enabled) ?
+          <canvas data-testid="cameraimage-canvas" height={scaledHeight} ref={canvasRef} width={scaledWidth} />
+          : <div class="text-center text-green-500">Camera is disabled in config, no stream or snapshot available!</div>
```
I think leaving this as white text would be better.

Not 100% sure if this is how we will want this to look in the end (since we are also going to address the situation where a camera has an error to replace the "green screen") but it's probably good for now until it is addressed in a separate PR.

banthungprong committed:
changed color to white, added top padding in disabled camera info

banthungprong committed:
changed indentation

banthungprong commented:
A few lines above the indentation was used in the same way. It's not really consistent in the complete file.

NickM-27 commented:
> A few lines above the indentation was used in the same way. It's not really consistent in the complete file.
> But thanks for the replies and reviews. It's really helpful and nice to work like this.

The ones that are indented start with `WARNING:` or `NOTE:` and are one statement. General descriptions around how the settings work should not be indented.

Zaschii commented:
Is there also the possibility to disable a cam per mqtt?
Because i have a webcam which is not always online - so how can i automate this (changing the config file per script is not really a smooth solution in my mind). Thx & BR

TheSleepySlee commented:
This works perfectly, thanks for this new option!
